natss/pkg/dispatcher/dispatcher.go

```diff
@@ -318,23 +318,43 @@ func (s *SubscriptionsSupervisor) unsubscribe(channel provisioners.ChannelRefere
 	return nil
 }
 
-func getSubject(channel provisioners.ChannelReference) string {
+func getSubject(channel channels.ChannelReference) string {
 	return channel.Name + "." + channel.Namespace
 }
 
-func (s *SubscriptionsSupervisor) getHostToChannelMap() map[string]provisioners.ChannelReference {
-	return s.hostToChannelMap.Load().(map[string]provisioners.ChannelReference)
+func (s *SubscriptionsSupervisor) getHostToChannelMap() map[string]channels.ChannelReference {
+	return s.hostToChannelMap.Load().(map[string]channels.ChannelReference)
 }
 
-func (s *SubscriptionsSupervisor) setHostToChannelMap(hcMap map[string]provisioners.ChannelReference) {
+func (s *SubscriptionsSupervisor) setHostToChannelMap(hcMap map[string]channels.ChannelReference) {
 	s.hostToChannelMap.Store(hcMap)
 }
 
+// NewHostNameToChannelRefMap parses each channel from cList and creates a map[string(Status.Address.HostName)]ChannelReference
+func NewHostNameToChannelRefMap(cList []messagingv1alpha1.Channel) (map[string]channels.ChannelReference, error) {
+	hostToChanMap := make(map[string]channels.ChannelReference, len(cList))
+	for _, c := range cList {
+		url := c.Status.Address.GetURL()
+		if cr, ok := hostToChanMap[url.Host]; ok {
+			return nil, fmt.Errorf(
+				"duplicate hostName found. Each channel must have a unique host header. HostName:%s, channel:%s.%s, channel:%s.%s",
+				url.Host,
+				c.Namespace,
+				c.Name,
+				cr.Namespace,
+				cr.Name)
+		}
+		hostToChanMap[url.Host] = channels.ChannelReference{Name: c.Name, Namespace: c.Namespace}
+	}
+	return hostToChanMap, nil
+}
+
+
 // UpdateHostToChannelMap will be called from the controller that watches natss channels.
 // It will update internal hostToChannelMap which is used to resolve the hostHeader of the
 // incoming request to the correct ChannelReference in the receiver function.
-func (s *SubscriptionsSupervisor) UpdateHostToChannelMap(ctx context.Context, chanList []eventingv1alpha1.Channel) error {
-	hostToChanMap, err := provisioners.NewHostNameToChannelRefMap(chanList)
+func (s *SubscriptionsSupervisor) UpdateHostToChannelMap(ctx context.Context, chanList []messagingv1alpha1.Channel) error {
```

Comment:
It feels odd to depend on messagingv1alpha1.Channel here. Should this be NatssChannel?